### PR TITLE
Partykit fix

### PR DESCRIPTION
**Pull request by bblodfon**
Solves https://github.com/mlr-org/mlr3extralearners/issues/327

### CODE REVIEW

File: tests/testthat/test_partykit_surv_cforest.R [hunk 1]
```diff
@@ -1,6 +1,20 @@
 test_that("autotest", {
-  learner = LearnerSurvCForest$new()
+  set.seed(1)
+  learner = lrn("surv.cforest", ntree = 50)
```
[the remark this answers is not in the document]
Comment (bblodfon) [reply tests/testthat/test_partykit_surv_cforest.R:3]: Ah I can disable it, if that's okay? => ie do `ntree = 2` and `exclude = 'sanity'` and it works (no need to set the seed)

[the remark this answers is not in the document]
Comment (bblodfon) [reply tests/testthat/test_partykit_surv_cforest.R]: just disabling it is okay, I updated the test code

**Comment by bblodfon** [comment]
@RaphaelS1 quick check here, what we discussed, didn't found any issues myself

File: R/learner_partykit_surv_cforest.R [hunk 2]
```diff
@@ -133,21 +133,17 @@ LearnerSurvCForest = R6Class("LearnerSurvCForest",
...
-        time = c(0, z$time, max(z$time) + 1e-3)
-        surv = c(1, z$surv, 0)
-        data.frame(x = time, cdf = 1 - surv)
+      times = lapply(preds, function(p) p$time)
+      utimes = sort(unique(unlist(times)))
+      res = lapply(preds, function(p) {
+        # p is a `survfit` object
+        cdf = matrix(data = 1 - p$surv, ncol = 1) # 1 observation (column), rows => times
+        # extend cdf to 'utimes', return survival
+        distr6:::C_Vec_WeightedDiscreteCdf(utimes, p$time, cdf = cdf, FALSE, FALSE)
```
[the remark this answers is not in the document]
Comment (bblodfon) [reply R/learner_partykit_surv_cforest.R:142]: to be updated

File: NEWS.md [hunk 3]
```diff
@@ -1,6 +1,7 @@
 # mlr3extralearners 0.7.1-9000
 
-# Added method `selected_features()` to CoxBoost survival learners (thanks to @bblodfon)
+* Fix: Replace hardcoded `VectorDistribution`s from partykit survival learners with survival matrices (`Matdist`) (thanks to @bblodfon)
```
Comment (sebffischer) [inline NEWS.md:3]: how is this a fix? Does this fix a bug?

File: R/learner_partykit_surv_cforest.R [hunk 4]
```diff
@@ -133,21 +133,20 @@ LearnerSurvCForest = R6Class("LearnerSurvCForest",
...
-      x = lapply(preds, function(z) {
-        time = c(0, z$time, max(z$time) + 1e-3)
-        surv = c(1, z$surv, 0)
-        data.frame(x = time, cdf = 1 - surv)
-      })
-      distr = distr6::VectorDistribution$new(
-        distribution = "WeightedDiscrete",
-        params       = x,
-        decorators   = c("CoreStatistics", "ExoticStatistics"))
+      times = lapply(preds, function(p) p$time)
```
Comment (sebffischer) [inline R/learner_partykit_surv_cforest.R:136]: map_dbl

File: R/learner_partykit_surv_cforest.R [hunk 5]
```diff
@@ -133,21 +133,20 @@ LearnerSurvCForest = R6Class("LearnerSurvCForest",
...
+      res = lapply(preds, function(p) {
+        # p is a `survfit` object
+        cdf = matrix(data = 1 - p$surv, ncol = 1) # 1 observation (column), rows => times
+        # extend cdf to 'utimes', return survival
+        extend_times(utimes, p$time, cdf = cdf, FALSE, FALSE)
+      })
+      surv = do.call(cbind, res) # rows => times, columns => obs
 
-      list(crank = crank, distr = distr)
+      .surv_return(times = utimes, surv = t(surv))
```
Comment (sebffischer) [inline R/learner_partykit_surv_cforest.R:149]: if you now transpose surv, why don't you create it the right way in the first place?

**Review by sebffischer — CHANGES_REQUESTED** [review]

File: NEWS.md (hunk 3 above)
Comment (bblodfon) [reply NEWS.md:3]: optimization? Refactoring?

Comment (sebffischer) [reply NEWS.md:3]: I am using this as my commit template: 


# ----------------------------------------------------------
# Header - type(scope): Brief description
# ----------------------------------------------------------
#    * feat             A new feature - SemVar PATCH
#    * fix              A bug fix - SemVar MINOR
#    * BREAKING CHANGE  Breaking API change - SemVar MAJOR
#    * docs             Change to documentation only
#    * style            Change to style (whitespace, etc.)
#    * refactor         Change not related to a bug or feat
#    * perf             Change that affects performance
#    * test             Change that adds/modifies tests
#    * build            Change to build system
#    * ci               Change to CI pipeline/workflow
#    * chore            General tooling/config/min refactor
# ----------------------------------------------------------


# ----------------------------------------------------------
# Body - More description, if necessary
# ----------------------------------------------------------
#    * Motivation behind changes, more detail into how
#      functionality might be affected, etc.
# ----------------------------------------------------------


# ----------------------------------------------------------
# Footer - Associated issues, PRs, etc.
# ----------------------------------------------------------
#    * Ex: Resolves Issue #207, see PR #15, etc.
# -------------------------------------------------------

File: R/learner_partykit_surv_cforest.R (hunk 4 above)
Comment (bblodfon) [reply R/learner_partykit_surv_cforest.R:136]: Can't get it to work! `preds` is a list of objects for which you want to get the `$time` vector. using `map_dbl`:

Comment (bblodfon) [reply R/learner_partykit_surv_cforest.R:136]: image

File: R/learner_partykit_surv_cforest.R (hunk 5 above)
Comment (bblodfon) [reply R/learner_partykit_surv_cforest.R:149]: It's the output of the `C_Vec_WeightedDiscreteCdf` (`extend_times` now) that has the rows and columns reversed and in the for loop above you get each row of the end matrix reversed via this function. Which means that you either do a `t` transpose on each observation (making a column to a row) after calling `extend_times`, or what I did (leave all columns, `cbind` them, transpose at the end) => second is faster was my thought

File: R/learner_partykit_surv_cforest.R (hunk 4 above)
Comment (sebffischer) [reply R/learner_partykit_surv_cforest.R:136]: Ah sorry yes this is what I thought

File: R/learner_partykit_surv_cforest.R (hunk 5 above)
Comment (sebffischer) [reply R/learner_partykit_surv_cforest.R:149]: makes sense!

**Review by sebffischer — APPROVED** [review]

**Pull request merged**